include/aws/common/linked_list.h

```diff
@@ -37,28 +37,100 @@ struct aws_linked_list {
     struct aws_linked_list_node tail;
 };
 
+/**
+ * These functions need to be defined first as they are used in pre
+ * and post conditions.
+ */
+
+/**
+ * Tests if the list is empty.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_empty(const struct aws_linked_list *list) {
+    return list->head.next == &list->tail;
+}
+
+/**
+ * Checks that a linked list is valid.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_is_valid(const struct aws_linked_list *list) {
+    return list && list->head.next && list->head.prev == NULL && list->tail.prev && list->tail.next == NULL;
+}
+
+/**
+ * Checks that the prev of the next pointer of a node points to the node.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_node_next_is_valid(const struct aws_linked_list_node *node) {
+    return node && node->next && node->next->prev == node;
+}
+
+/**
+ * Checks that the next of the prev pointer of a node points to the node.
+ */
+AWS_STATIC_IMPL bool aws_linked_list_node_prev_is_valid(const struct aws_linked_list_node *node) {
```

Comment:
How do we make sure this is never called on the head?

Reply:
We cannot make sure, but I added a comment to explain the intent of this function, and the fact that this should not return true for head (and `next` for tail accordingly).